### PR TITLE
Add 3d plots to workbench

### LINKED ISSUE #21720: Add basic 3d plotting to workbench
Covers adding basic plotting to the workbench. 
This should include writing wrapper functions to allow 3d plots to be generated from Mantid workspaces as well as updating/adding the mantid projection.

### PR DESCRIPTION
**Pull request by ElliotAOram**
* Refactor of the Mantid Projection (`__init__.py` and `functions.py`) into three classes (to also include 3D projection.
* White space and formatting changes
* Exposed functions:
  * plot
  * scatter
  * plot_wireframe
  * plot_surface
  * contour
  * contourf

* Question to tester: do we want plot to be able to take a range of specNum? as far as I can see 2d plotting only allows one?

**To test:**

* use a mantid workspace as input to the 3D plotting functions in the workbench. Below is a tester script using a file from the `usage data`.

```
# The following line helps with future compatibility with Python 3
# print must now be used as a function, e.g print('Hello','World')
from __future__ import (absolute_import, division, print_function, unicode_literals)

# import mantid algorithms, numpy and matplotlib
from mantid.simpleapi import *

import matplotlib.pyplot as plt

import numpy as np

ws = Load("MUSR00015189")
ws = mtd['ws_1']
ws = Rebin(ws, "0,0.25, 32") 

fig = plt.figure()
ax = fig.add_subplot(111,projection='mantid3d')

# Plot a basic surface.
ax.plot(ws, specNum=1)
#ax.set_xlabel("Time")
#ax.set_zlabel("Counts")
#ax.set_ylabel("Spectrum Number")

plt.show()
```

* Play around and try to break the plotting. 

Fixes #21720 

**Release Notes** 
*Does not need to be in the release notes.*

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by AndreiSavici — CHANGES_REQUESTED** [review]
Need documentation (`mantid/docs/source/api/python/mantid/plots/index.rst`) and tests (`mantid/Framework/PythonInterface/test/python/mantid/plot`)

**Comment by AndreiSavici** [comment]
I would not add the plotting of a range of spectra. One can just add multiple lines. It is more in line with what matplotlib does. Also, that way you have better control on symbols, colors, linestyles, labels, etc.

File: Framework/PythonInterface/mantid/plots/__init__.py [hunk 1]
```diff
@@ -323,14 +320,184 @@ def tricontourf(self,*args,**kwargs):
...
+            fig, ax = plt.subplots(subplot_kw={'projection':'mantid3d'})
+            ax.plot(workspace) #for workspaces
+            ax.plot(x,y,z)     #for arrays
+            fig.show()
+
+        For keywords related to workspaces, see :func:`mantid.plots.functions.plot3D`
+        '''
+        if mantid.plots.helperfunctions.validate_args(*args):
+            mantid.kernel.logger.debug('using mantid.plots.plotfunctions3D')
+            return mantid.plots.plotfunctions3D.plot3d(self, *args, **kwargs)
```
Comment (AndreiSavici) [inline Framework/PythonInterface/mantid/plots/__init__.py]: For consistency with matplotlib, this function should call `mantid.plots.plotfunctions3D.plot`, so you should rename your `plot3d`

**Review by AndreiSavici — APPROVED** [review]

**Comment by AndreiSavici** [comment]
There is a missing section in the autogenerated documentation. I will add that as part of a different PR. The failing test is unrelated (I hope).

**Comment by martyngigg** [comment]
We've got some unstable tests from this unfortunately: http://builds.mantidproject.org/job/pull_requests-ubuntu-python3/7416/console

**Comment by martyngigg** [comment]
I think the `matplotlib.use("AGG")` needs to come *before* importing pyplot as it was in the tests before. @ElliotAOram Could you make a quick PR to fix that?

**Comment by AndreiSavici** [comment]
I'm fixing this as part of #21790

**Pull request merged**